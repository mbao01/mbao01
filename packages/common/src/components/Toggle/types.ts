import * as TogglePrimitive from "@radix-ui/react-toggle";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";
import { getToggleClasses } from "./constants";

export type BaseToggleProps = Omit<VariantProps<typeof getButtonClasses>, "variant" | "isLoading"> &
  VariantProps<typeof getToggleClasses>;

export type ToggleProps = React.ComponentPropsWithoutRef<typeof TogglePrimitive.Root> &
  BaseToggleProps;
