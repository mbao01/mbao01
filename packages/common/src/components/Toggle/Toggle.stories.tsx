import type { Meta, StoryObj } from "@storybook/react";
import { FontBoldIcon } from "@radix-ui/react-icons";
import { Toggle } from "./Toggle";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Toggle,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Toggle>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    variant: "default",
    children: "Toggle",
  },
};

export const IconContent: Story = {
  args: {
    variant: "default",
  },
  render: () => {
    return (
      <Toggle aria-label="Toggle italic">
        <FontBoldIcon className="h-4 w-4" />
      </Toggle>
    );
  },
};

export const PrimaryToggle: Story = {
  args: {
    variant: "primary",
    children: "Toggle",
  },
};

export const WideToggle: Story = {
  args: {
    wide: true,
    children: "Wide Toggle",
  },
};

export const TinyToggle: Story = {
  args: {
    size: "xs",
    children: "Tiny",
  },
};

export const OutlineToggle: Story = {
  args: {
    outline: true,
    children: "Click me!",
  },
};

export const DisabledToggle: Story = {
  args: {
    disabled: true,
    children: "Click me?",
  },
};

export const BlockToggle: Story = {
  args: {
    disabled: true,
    className: "btn-block",
    children: "Block Toggle",
  },
};
