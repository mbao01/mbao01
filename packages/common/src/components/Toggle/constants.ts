import { cva } from "../../libs";

export const getToggleClasses = cva(
  "transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost",
  {
    variants: {
      variant: {
        accent: "data-[state=on]:btn-accent",
        default: "data-[state=on]:btn-default",
        error: "data-[state=on]:btn-error",
        info: "data-[state=on]:btn-info",
        link: "data-[state=on]:btn-link",
        neutral: "data-[state=on]:btn-neutral",
        primary: "data-[state=on]:btn-primary",
        secondary: "data-[state=on]:btn-secondary",
        success: "data-[state=on]:btn-success",
        warning: "data-[state=on]:btn-warning",
      },
    },
  }
);
