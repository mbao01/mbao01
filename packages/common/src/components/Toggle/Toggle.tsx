"use client";

import * as React from "react";
import * as TogglePrimitive from "@radix-ui/react-toggle";
import { type ToggleProps } from "./types";
import { cn } from "../../utilities";
import { getToggleClasses } from "./constants";
import { getButtonClasses } from "../Button/constants";

const Toggle = React.forwardRef<
  React.ElementRef<typeof TogglePrimitive.Root>,
  ToggleProps
>(({ className, variant = "default", size, wide, outline, ...props }, ref) => (
  <TogglePrimitive.Root
    ref={ref}
    className={cn(
      getButtonClasses({ size, wide, outline }),
      getToggleClasses({ variant }),
      className
    )}
    {...props}
  />
));

Toggle.displayName = TogglePrimitive.Root.displayName;

export { Toggle };
