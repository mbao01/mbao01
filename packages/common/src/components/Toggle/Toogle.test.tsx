import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Toggle } from "./Toggle";

describe("Toggle", () => {
  it("renders a toggle", () => {
    const { asFragment } = render(<Toggle>Toggle button</Toggle>);

    expect(
      screen.getByRole("button", { name: "Toggle button" })
    ).toHaveAttribute("data-state", "off");
    expect(asFragment()).toMatchSnapshot();
  });

  it("should toggle button state", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(<Toggle>Toggle</Toggle>);
    const toggle = screen.getByRole("button", { name: "Toggle" });
    expect(toggle).toHaveAttribute("data-state", "off");

    await user.click(toggle);

    expect(toggle).toHaveAttribute("data-state", "on");

    expect(asFragment()).toMatchSnapshot();
  });
});
