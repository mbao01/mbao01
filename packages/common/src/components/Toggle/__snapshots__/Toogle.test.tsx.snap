// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > renders a toggle 1`] = `
<DocumentFragment>
  <button
    aria-pressed="false"
    class="btn min-h-fit h-10 transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost data-[state=on]:btn-default"
    data-state="off"
    type="button"
  >
    Toggle button
  </button>
</DocumentFragment>
`;

exports[`Toggle > should toggle button state 1`] = `
<DocumentFragment>
  <button
    aria-pressed="true"
    class="btn min-h-fit h-10 transition-colors disabled:pointer-events-none disabled:opacity-50 data-[state=off]:btn-ghost data-[state=on]:btn-default"
    data-state="on"
    type="button"
  >
    Toggle
  </button>
</DocumentFragment>
`;
